### PR TITLE
add test for timers

### PR DESCRIPTION
**Pull request by dirk-thomas**
http://ci.ros2.org/job/ros2_batch_ci_linux/133/
http://ci.ros2.org/job/ros2_batch_ci_osx/65/
http://ci.ros2.org/job/ros2_batch_ci_windows/92/

### CODE REVIEW

**Comment by wjwwood** [comment]
You'll have to use spin or spin_once to test what I'm talking about. Spin some will result in a non-blocking spin, so timers could be handled when rmw_wait wakes up. You need a blocking call to rmw_wait or a timed call to rmw_wait in order to test what I was describing. That's why I suggested calling spin in a thread do it can block.

**Comment by dirk-thomas** [comment]
I have added the test you describe in 51d1565988c26fc718b2d3de0e21ca2aa1497e4b.

File: test_rclcpp/test/test_timer.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+  auto shutdown =
+    []() -> void
+    {
+      rclcpp::shutdown();
+      std::cout << "shutdown()" << std::endl;
+    };
+
+  auto shutdown_timer = node->create_wall_timer(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(4.5 * period), shutdown);
```
Comment (wjwwood) [inline test_rclcpp/test/test_timer.cpp]: Using the mechanism you're testing in order to shutdown the test is not a reliable way to make sure the test stops. In fact if this test fails as I expect it to, this callback will not be executed, the test will hang, and the test runner will have to kill it.

File: test_rclcpp/test/test_timer.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  auto spinner =
+    [&executor, &node]() -> void
+    {
+      std::cout << "spin() until shutdown timer" << std::endl;
+      executor.add_node(node, false);
+      executor.spin();
+    };
+
+  auto thread = std::thread(spinner);
+  thread.join();
```
Comment (wjwwood) [inline test_rclcpp/test/test_timer.cpp]: Why not just drop the shutdown timer, then do this:

```c++
auto thread = std::thread(spinner);
std::this_thread::sleep_for(period * 4);
rclcpp::shutdown();
thread.join();
```

Comment (dirk-thomas) [inline test_rclcpp/test/test_timer.cpp]: Updated the test in 78fcfdba257c38ba5d8a537b8e0b1c1254bb2a70

**Comment by dirk-thomas** [comment]
http://ci.ros2.org/job/ros2_batch_ci_linux/135/
http://ci.ros2.org/job/ros2_batch_ci_osx/67/
http://ci.ros2.org/job/ros2_batch_ci_windows/94/

**Comment by wjwwood** [comment]
lgtm, thanks for iterating on it.

**Comment by dirk-thomas** [comment]
With the increased timeout (9c2114d96ba6893dcf3b4180ea6d085fa6bd6c38) the tests pass:
http://ci.ros2.org/job/ros2_batch_ci_linux/136/
http://ci.ros2.org/job/ros2_batch_ci_osx/68/

I have also created ament/ament_cmake#19 to better handle this case.

**Pull request merged**